### PR TITLE
msvc-fix

### LINKED ISSUE #3: Need to link "shlwapi.lib" when building for Windows
Hi again! :)

Just tried to build & run my benchmarks (linking `benchmark`) on MSVC and got this:

```
benchmark.lib(sysinfo.lib.obj) : error LNK2019: unresolved external symbol __imp_SHGetValueA referenced in function "double __cdecl benchmark::`anonymous namespace'::GetCPUCyclesPerSecond(enum benchmark::CPUInfo::Scaling)" (?GetCPUCyclesPerSecond@?A0x511b6745@benchmark@@YANW4Scaling@CPUInfo@2@@Z)
```

Adding this when building makes it work (mentioned [here](https://github.com/google/benchmark/blob/847c0069021ade355b7678a305f3ac4e4d6f7e79/README.md#windows-with-cmake)):
`config.cxx.loptions+="/link shlwapi.lib"`

So I guess this should go along with the exported linker flags when building for Windows.

### PR DESCRIPTION
**Pull request by Swat-SomeBug**
Working implementation of on msvc.
To merge in:

- [X] ~~Builds with MSVC with default configuration (`config.bin.lib=both`)~~
* Even though the library itself can be built as a shared lib, several features of the library cannot be used since `upstream` doesn't support it. 
- [X] ~~Builds when `config.bin.lib=shared`~~
* Seems to be a problem with [upstream](https://github.com/google/benchmark) regarding support of shared libs on windows reported under: google/benchmark#640.
* Confirm: Same issue as above. Shared lib is not supported by `upstream` with msvc.
* Shared libs with `msvc` is unsupported. Confirmed to be the same root issue as in the link above.
- [X] Builds when `config.bin.lib=static`
- [X] All tests pass
* Currently `tests/basics/exe{reporter_output_test}` fails with non-zero error code. This is general problem even with `upstream` on windows due to incorrect path separator. However, `ctest` for some reason doesn't detect this and flags as a pass.
* `tests/basics/exe{reporter_output_test}` is disabled on `windows`

### CODE REVIEW

**Comment by helmesjo** [comment]
Hmm I can see you add it to `config.export.libs` but I still see this:
`benchmark.lib(sysinfo.lib.obj) : error LNK2019: unresolved external symbol __imp_SHGetValueA referenced in function "private: __cdecl benchmark::CPUInfo::CPUInfo(void)" (??0CPUInfo@benchmark@@AEAA@XZ)`

Again, I'm not too familiar with build2, but should it perhaps be `cxx.export.loptions`? Or what's the difference between `libs` and `loptions`? Locally it works regardless if I add it to `config.cxx.loptions` or `config.cxx.libs`...

**Edit**
To answer myself since I got curious, [GNU says this](https://www.gnu.org/software/make/manual/html_node/Implicit-Variables.html):
> **LDFLAGS**
Extra flags to give to compilers when they are supposed to invoke the linker, ‘ld’, such as -L. Libraries (-lfoo) should be added to the LDLIBS variable instead.
**LDLIBS**
Library flags or names given to compilers when they are supposed to invoke the linker, ‘ld’. LOADLIBES is a deprecated (but still supported) alternative to LDLIBS. Non-library linker flags, such as -L, should go in the LDFLAGS variable.

**Comment by helmesjo** [comment]
I tried making your changes manually (in my configuration) and it worked as it should, which made it obvious that I've missed some step (I only changed my `repositories.manifest` to point to this new branch, though `bdep update` did indicate that `google-benchmark` was fetched). Files in my configuration still didn't include these changes.

I then did `bdep sync` & `bdep fetch` (tried both orders) and it did update the `src/buildfile` under `myconfig/.bdep/repos/...`, but _not_ the `src/buldfile` under `myconfig/google-benchmark...`, which is the one used when I build my project.

In short, your changes work fine but either I'm missing some build2 command after modifying the `repositories.manifest` (probably), or build2 doesn't do the correct thing when just swapping which branch a prerequisite points to.

I've scimmed through the documentation a few times but can't find the secret sauce.

@boris-kolpackov Sorry for bringing you in again (and going a bit OT for this PR), but could you shine some light on this?

**Comment by Swat-SomeBug** [comment]
@helmesjo @boris-kolpackov I’m able to reproduce the above mentioned issue. I did the following in a my simple testing:

* Add the git repo with `master` branch in `repositories.manifest`
* Initialize the test project with `bdep init`
* Build
* Change the branch to `msvc-fix`
* Build

The way I solved it was to `deinit` the project after switching the branch. This causes `google-benchmark` to be dropped. Then perform `bdep init && b` which causes the right commit to be fetched and built.

I believe the issue is not related to the PR in general.
@boris-kolpackov is such a use case valid/supported by `build2`?

**Comment by helmesjo** [comment]
Okay, so a third scimming brought me to section **2.3** in the [`Toolchain introduction`](https://build2.org/build2-toolchain/doc/build2-toolchain-intro.xhtml#guide-add-remove-deps):

> To get rid of a dependency, we simply remove it from the manifest file and synchronize the project. For example, assuming libhello is no longer mentioned as a dependency in our manifests

I removed `google-benchmark` from my `manifest`, synched (it got dropped), added it again, synched, and it got the latest with these changes (note that `google-benchmark` is not version constrained in my `manifest`, since the package is still a pre-release, which I assume is the root cause of my issues).

**Comment by Swat-SomeBug** [comment]
> (note that google-benchmark is not version constrained in my manifest, since the package is still a pre-release, which I assume is the root cause of my issues)


@helmesjo by the way, `build2` supports pre-release versioning. For me that’s one of `build2`’s strong points compared to other build systems. If really required, one could pin down a specific pre-release version based on commit id 😉

**Comment by helmesjo** [comment]
@Swat-SomeBug Yeah I agree, though I guess I should've been a bit more specific. What I can't seem to do is just specify:
`depends: google-benchmark ^1.6.0` (bdep spits out errors, as expected)
so I just have:
`depends: google-benchmark`

And as per usual, while writing this I found the answer:
> The shortcut operators can only be used with standard versions (a semantic version without the pre-release part is a standard version). They are equivalent to the following ranges. **The X.Y.Z- version signifies the earliest pre-release in the X.Y.Z series; see Package Version for details.**

So adding the trailing `-` made build2 accept the constraint (though the above mentioned issue still remains, prerequisite must be dropped & re-added to point to the correct commit).

**Comment by helmesjo** [comment]
@Swat-SomeBug Anyways, sorry for going off-topic but I figured it was worth to be mentioned. The changes themselves work fine for me, so feel free to complete!

**Comment by Swat-SomeBug** [comment]
To get this PR to be merged in:

- [X] Builds with MSVC with default configuration (`config.bin.lib=both`)
- [ ] Builds when `config.bin.lib=shared`
* Seems to be a problem with upstream regarding shared libs reported [here](https://github.com/google/benchmark/blob/847c0069021ade355b7678a305f3ac4e4d6f7e79/README.md#windows-with-cmake).
* Currently investigating to see if the issue is exactly the same.
- [X] Builds when `config.bin.lib=static`
- [ ] All tests pass
* Currently tests/basic/exe{diagnostics_test} fails with non-zero error code. Suspecting if this may have to do with exception flags.

**Comment by boris-kolpackov** [comment]
> I then did `bdep sync` & `bdep fetch` (tried both orders) and it did update the `src/buildfile` under `myconfig/.bdep/repos/...`, but not the `src/buldfile` under `myconfig/google-benchmark...`, which is the one used when I build my project.

> In short, your changes work fine but either I'm missing some `build2` command after modifying the repositories.manifest (probably), or `build2` doesn't do the correct thing when just swapping which branch a prerequisite points to.

> I've scimmed through the documentation a few times but can't find the secret sauce.

@helmesjo If I understand the setup correctly, `google-benchmark` is a dependency of your project. In this case, I believe the behavior is as expected: `build2` will only upgrade a dependency if necessary (e.g., to satisfy a new/changed version constraint) or if explicitly requested by the user with `bdep sync --upgrade|-u`. For example, `bdep sync -fur` will re-fetch all the repository information (`-f`), and upgrade all the dependencies (`-u`), recursively (`-r`). There are some other relevant options (e.g., `--patch|-p`) which are all discussed in the `bdep-sync(1)` man page.

**Comment by helmesjo** [comment]
@boris-kolpackov Thank you for the explanation. `bdep sync -fur` indeed did what I wanted.
>`build2` will only upgrade a dependency if necessary (e.g., to satisfy a new/changed version constraint) or if explicitly requested by the user with `bdep sync --upgrade|-u`.

Would it not make sense to _also_ upgrade a dependency if it's _location_ changes (in `repositories.manifest`)? Eg. by changing which branch it points to, or if it suddenly point to a completely different repository? I assume `build2` keeps track of where a prerequisite originated from (which repository). I'm afraid this would otherwise risk a lot of confusion by `build2` users, mainly because `bdep update` _does_ fetch the new repository but doesn't use it (so the output seemingly tricked me a bit).
The alternative would be that it at least suggest the user to do a `bdep sync ...` to achieve what they (most likely) want.

Anyways, not related to this PR so I guess this could be continued elsewhere. Thanks again!

File: build/root.build [hunk 1]
```diff
@@ -8,3 +8,9 @@ cxx{*}: extension = cc
 # The test target for cross-testing (running tests under Wine, etc).
 #
 test.target = $cxx.target
+
+# Disable support for building shared libs on windows with msvc
+# Upstream does not support this
+# Public headers are not marked appropriately with declspec(dllimport) and declspec(dllexport)
+# Causes severe problems with linking due to static variables and inline functions in headers
+assert! ($config.bin.lib != 'static' && $cxx.target.system == 'win32-msvc') 'Google Benchmark does not support building shared libs with MSVC. Explicitly set config.bin.lib=static when using MSVC'
```
Comment (helmesjo) [inline build/root.build]: Correct me if I'm wrong, but won't this force dependent projects to also be built as `static`? Or can one specify to only build `google-benchmark` as static?

`bdep init config.bin.lib=static` still prints this error though (adding a `print` shows this value twice, and it's `both` first time, `static` second). It also fails when doing `bdep deinit` (since it appears to check this then as well). Anyways, I don't think this sort of hard enforcement is the correct way to go about this (can't find it now but I remember reading it somewhere in the build2 docs/issues). Also, as mentioned [here](https://github.com/build2-packaging/google-benchmark/issues/3#issuecomment-1006137466), it worked fine before by simply importing the static target (`google-benchmark%liba{benchmark}`).

Then again, it would be confusing to build `both` and only get the static `benchmark` target...

Comment (Swat-SomeBug) [reply build/root.build]: As I understand all `build2` configs are amalgamations. And during `init`, one can pass package specific options.

Importing `liba{}` vs `libs{}` vs `lib{}` is consumer’s choice unless there is very special underlying reason. Consumers typically would want to use `lib{}` unless only `liba{}` is available, for example `liba{benchmark-main}`.

So options are:
* Support `liba{benchmark}` only, for all platforms and compilers because one specific combination doesn’t support shared libs.
* Disable build on the single combination and ask the user to pass the `config` option.

I would like to support `libs{}` with gcc, clang and mingw on windows, so can’t enforce building only `liba{benchmark}`.

The config option can be for the whole amalgamation or for this specific package. This is again the consumers choice. But the general idea is that, this choice only needs be made for only one specific platform and compiler combination.

Comment (Swat-SomeBug) [reply build/root.build]: > bdep init config.bin.lib=static still prints this error though (adding a print shows this value twice, and it's both first time, static second).

On a de-initialised configuration could you try the following command
```
bdep init { config.bin.lib=static }+ "?google-benchmark"
```
The extra `{ }+` is for argument grouping and applies only to `google-benchmark` while using default configuration for all other packages.

Edit:
The command without `{ }+` requests `bdep` to configure only the primary/main package with this option. In principle this is ok (not something related to `google-benchmark`), but at the same time, `config.bin.lib=both` is specified for `google-benchmark`, which is not ok with MSVC. Because the configuration is an amalgamation, if you want to build only static libs for all packages, one could add `config.bin.lib=static` to the amalgamation's `config.build` to enforce this. My point is, `build2` offers the flexibility to build different packages with different options. At the same time, one can enforce a common set of options to all packages within a configuration.

For `google-benchmark` specifically, because `upstream` does not (unfortunately) support shared libs with MSVC, we must ensure consumers don't build it.
TBH for local testing, I was able to build the package as a share lib without problems. As long as one uses only the core functionality and never refers to any inline functions or static variables, everything works fine. This however, cannot be enforced.

Comment (helmesjo) [reply build/root.build]: > As I understand all `build2` configs are amalgamations. And during `init`, **one can pass package specific options**.

How? I can't find anything about this in the documentation. I'd expect something like `config.benchmark.bin.lib=static`. And still, `config.bin.lib=static` didn't work either.

> Importing `liba{}` vs `libs{}` vs `lib{}` is consumer’s choice unless there is very special underlying reason. Consumers typically would want to use `lib{}` unless there `liba{}` only export for example `liba{benchmark-main}`.
> I would like to support `libs{}` with gcc and clang and mingw, so can’t enforce building only `liba{benchmark}`.
> 

I agree that `lib{}` should be used for other platforms. I rather thought something like `if msvc -> setup liba{benchmark} only` (so you get an error about missing `benchmark` target if using `lib{benchmark}` and building shared, or explicitly `libs{benchmark}`).

Wouldn't it be better to not have any enforcement at all, perhaps only a warning referring to [this](https://github.com/google/benchmark/issues/639)?

Or... Implement your own version of [`WINDOWS_EXPORT_ALL_SYMBOLS`](https://cmake.org/cmake/help/latest/prop_tgt/WINDOWS_EXPORT_ALL_SYMBOLS.html) (better yet if it was available in `build2`). :)

Comment (helmesjo) [reply build/root.build]: > On a de-initialised configuration could you try the following command
> 
> ```
> bdep init { config.bin.lib=static }+ "?google-benchmark"
> ```

Yeah... Neither cmd nor powershell is swallowing that command. :)

**Comment by boris-kolpackov** [comment]
> Would it not make sense to also upgrade a dependency if it's location changes (in `repositories.manifest`)? Eg. by changing which branch it points to, or if it suddenly point to a completely different repository? I assume `build2` keeps track of where a prerequisite originated from (which repository). I'm afraid this would otherwise risk a lot of confusion by `build2` users, mainly because `bdep update` does fetch the new repository but doesn't use it (so the output seemingly tricked me a bit).
The alternative would be that it at least suggest the user to do a `bdep sync ...` to achieve what they (most likely) want.

@helmesjo, could you file an issue for this? It's a bit more complicated that what you've described (I will elaborate in the issue) but I think you have a good point, nevertheless.

File: build/root.build (hunk 1 above)
Comment (Swat-SomeBug) [reply build/root.build:16]: > Or... Implement your own version of WINDOWS_EXPORT_ALL_SYMBOLS (better yet if it was available in build2). :)

Comment (Swat-SomeBug) [reply build/root.build]: > Or... Implement your own version of WINDOWS_EXPORT_ALL_SYMBOLS (better yet if it was available in build2). :)

`build2` actually has this. Boris implemented this feature and is available starting 0.14.0.
As I mentioned in my comment, building a shared lib works fine. The under laying problem is not the export of the symbols. Rather, it's import. One commit prior in this PR, with ref: fcc669e, I already implemented symbol exports using `build2`'s `def{}` which works well and exports all symbols.
The issue is the recognition of inline and static symbols. MSVC by default will create multiple definition of inline and static symbols in every TU and `link.exe` does not load symbols from the dll when the definition is available in the TU, unless they are annotated as `__declspec(dllimport)`. This is also the root problem for `upstream`. 

Things actually got nasty in linking the dll lib with `tests/basics/obj{diagnostics_test}` which accessed an inline function. MSVC created 2 definitions of the exact same symbol. I verified this by dumping the symbol table in the TU as well as the dll. With gcc and clang when the linker encounters multiple identical definitions, it may choose to use one and discard others. The problem with MSVC and `link.exe` is that, symbols are not loaded into the symbol table when the definition is already available unless they are annotated correctly. So there is no link error.
In effect though, the function call is available at 2 separate memory locations storing 2 separate sets of variables (static variables in this case).

I actually patched several internal headers locally to avoid the above mentioned problem. In principle, we can patch `upstream`'s public header as well. But it starts making less and less sense for me in terms of maintenance effort. If `upstream` itself states not to support shared libs, I wonder if it makes sense for the package to support it.

Comment (Swat-SomeBug) [reply build/root.build]: > Wouldn't it be better to not have any enforcement at all, perhaps only a warning referring to this?

I went with the more conservative approach first.
It is better not to initiate the build at all when the library configuration is not as expected. An upfront `assert` can save quite a lot of time and effort rather than allowing consumers to mistakenly link against the shared lib.

To make it work properly, consumers would have to pay attention to a single log message in-between a wall of text during the build. In the end, depending on how individual benchmarks are written, they can have very weird linker issue, like the one I had. By default, `config.bin.exe.lib=shared static`, which means a shared lib is preferred to a static lib. So consumers may not even realise the underlaying problem. `link.exe` would happily create an executable and devs would redo the same investigation all over again when something goes wrong.

Comment (Swat-SomeBug) [reply build/root.build]: > Yeah... Neither cmd nor powershell is swallowing that command. :)

I think the problem is with escaping certain characters. I personally using git-bash so I got an error for only the `?google-benchmark` part. On `cmd` the same command worked for me. On `powershell` quotes around `{ config.bin.lib=static }+` worked:
```
bdep init "{ config.bin.lib=static }+" "?google-benchmark"
```
`config.build` can even be hand edited.

For exact syntax, have a look at
```
bdep help init
bdep help argument-grouping
```
I think the same is found at: [bdep-doc](https://build2.org/bdep/doc/bdep.xhtml) and [bdep-argument-grouping]().

Comment (helmesjo) [reply build/root.build]: > I think the problem is with escaping certain characters. I personally using git-bash so I got an error for only the `?google-benchmark` part. On `cmd` the same command worked for me. On `powershell` quotes around `{ config.bin.lib=static }+` worked:
> 
> ```
> bdep init "{ config.bin.lib=static }+" "?google-benchmark"
> ```

Not seeing the same thing here... I've tried all sorts of combinations, and it fails in all but one case with bash (see below).

**Powershell:**
Does not work:
```
bdep init "{ config.bin.lib=static }+" "?google-benchmark"
error: missing value for option '-n'
```
Does not work:
```
bdep init -C -@msvc cc config.cxx=cl.exe "{ config.bin.lib=static }+" "?google-benchmark"
error: expected variable assignment instead of '{ config.bin.lib=static }+'
  info: use double '--' to treat this argument as buildspec
```
**Bash:**
Works:
```
bdep init @msvc-bash { config.bin.lib=static }+ "?google-benchmark"
```
Does not work: 
```
bdep init @msvc-bash "{ config.bin.lib=static }+" "?google-benchmark"
```
Does not work:
```
bdep init -C @msvc-bash cc config.cxx=cl.exe { config.bin.lib=static }+ "?google-benchmark"
error: unexpected grouped argument 'config.bin.lib=static' for argument '?google-benchmark'
```

So only one case worked, and that requires me to first create the config (which will then error out on the assert), then deinit and init again.

> `config.build` can even be hand edited.

Again, can't achieve this. Adding this to `config.build` (quotes, no quotes, whatever):
`{ config.bin.lib=static }+ ?google-benchmark`
Gives:
```
bdep init -C -@msvc cc config.cxx=cl.exe
error: Google Benchmark does not support building shared libs with MSVC. Explicitly set config.bin.lib=static when using MSVC
```

> I think the same is found at: [bdep-doc](https://build2.org/bdep/doc/bdep.xhtml) and bdep-argument-grouping.

Thanks!

## **In summary**
So TBH, given all the hassle surrounding this (unless you find my error in how I setup `config.build` incorrectly), I'd vote for this:

Re-add `def{...}` & remove the one test that uses `PerfCounters::kSupported` (which is under the `internals` namespace anyways) & fails. I've tested it and all just works. No errors, no special configs for a certain platform etc. 
See my draft here (just `git apply` onto this branch): [Export-all-symbols-when-building-shared-for-windows.txt](https://github.com/build2-packaging/google-benchmark/files/7842197/Export-all-symbols-when-building-shared-for-windows.txt)

Comment (Swat-SomeBug) [reply build/root.build]: @helmesjo,
I'm unable to reproduce these errors.
At my end the commands work with various quoting on `git-bash`, `cmd`, `powershell`. Tried `bash` and `zsh` on linux as well.
In principle, there is a way to do it. The commands are documented. If you are facing issues with `bdep init` then it's probably a good idea to open an issue in the [build2/bdep](https://github.com/build2/bdep) repo. The initialisation syntax itself has nothing to do with this package. They are orthogonal.

Comment (Swat-SomeBug) [reply build/root.build]: > So only one case worked, and that requires me to first create the config (which will then error out on the assert), then deinit and init again.

This also what I experienced so far. One can add a feature request to allow both config creation + config passing in [build2/bdep](https://github.com/build2/bdep). Again this is not related to the package. The package only requires a certain configuration. How this configuration is achieved, like using 2 `bdep` commands instead of 1, or using `b config` form or using `bpkg config` is not relevant.

> Again, can't achieve this. Adding this to config.build (quotes, no quotes, whatever):
{ config.bin.lib=static }+ ?google-benchmark

This syntax should not be used when editing `config.build`. Please have a look at [build2 docs. Section 1.2](https://build2.org/build2/doc/build2-build-system-manual.xhtml#intro-proj-struct) for a general overview and project/configuration structure. For details about the `config` module have a look at [build2 docs. Section 1.4.1](https://build2.org/build2/doc/build2-build-system-manual.xhtml#intro-operations-config)
In summary, please place the variable `config.bin.lib=static` under `<your-configuration-path>/google-benchmark-<version-number>/build/config.build`.
Notice that, the configuration itself has a minimum of 3 `config.build` files.
* 1 under `<your-configuration-path>` -> This is a global configuration file applied to all packages within the config
* 1 under `<your-main-pkg>` -> Pkg specific `config.build`
* 1 under `google-benchmark-<versionnumber>` -> `google-benchmark` specific `config.build`

`config.build` can be hand made or generated. They are meant to be hand-edited post generation as well.
Again, configuration setup is not directly related to this package and is off topic.
I found this [video](https://youtu.be/hHRaZy1LZPA) to be quite useful in getting started with `build2`.

Comment (Swat-SomeBug) [reply build/root.build]: > Re-add def{...} & remove the one test that uses PerfCounters::kSupported (which is under the internals namespace anyways) & fails. I've tested it and all just works. No errors, no special configs for a certain platform etc.
See my draft here (just git apply onto this branch): Export-all-symbols-when-building-shared-for-windows.txt

As I said, building a shared lib is no problem. And for simple tests this is ok and works fine. Problem comes when consumers utilise more features offered by the library. By allowing `libs{benchmark}` to build, we create a false sense that everything works, when we already know it's not the right thing to do. This is not a good design practice in general for both `c++` code and `buildfile`s. Flagging an error and asking the user to correct an incorrect/unsupported configuration, in my opinion is a fair way to deal with it. This is not limited to a set of tests either. When the package precondition is violated, it does not make any sense to continue building `libs{benchmark}`. 

My ideal solution would be to disable `libs{benchmark}` with `msvc`, log a warning and continue building only `liba{benchmark}`. At consumer side, access it as `lib{benchmark}` and get `liba{benchmark}` with `msvc`, on all other platforms, get what is set by the user. If the consumer forces `config.bin.lib=shared` then error out under `msvc`. I don't know how this can be achieved currently. @boris-kolpackov any idea/suggestion on how to achieve this?

**Comment by helmesjo** [comment]
> @helmesjo, could you file an issue for this? It's a bit more complicated that what you've described (I will elaborate in the issue) but I think you have a good point, nevertheless.

@boris-kolpackov 
No problem! New issue added [here](https://github.com/build2/build2/issues/178).

File: build/root.build (hunk 1 above)
Comment (helmesjo) [reply build/root.build]: > As I said, building a shared lib is no problem. And for simple tests this is ok and works fine. Problem comes when consumers utilise more features offered by the library. By allowing `libs{benchmark}` to build, we create a false sense that everything works, when we already know it's not the right thing to do. This is not a good design practice in general for both `c++` code and `buildfile`s. Flagging an error and asking the user to correct an incorrect/unsupported configuration, in my opinion is a fair way to deal with it. This is not limited to a set of tests either. When the package precondition is violated, it does not make any sense to continue building `libs{benchmark}`.

My bad, I was under the impression that the only issue there was with building shared libs was one or two `static` variables (and I only saw "internal" ones used solely by tests). Thanks for clarifying.

> My ideal solution would be to disable `libs{benchmark}` with `msvc`, log a warning and continue building only `liba{benchmark}`. At consumer side, access it as `lib{benchmark}` and get `liba{benchmark}` with `msvc`, on all other platforms, get what is set by the user. If the consumer forces `config.bin.lib=shared` then error out under `msvc`.

Isn't this what you describe?
```
# root.build

if ($cxx.target.system == 'win32-msvc')
{
  if ($bin.lib == 'both')
  {
    warn 'Shared libs not supported for Windows, building static only.'
    bin.lib=static
  }
  elif($bin.lib == 'shared')
  {
    fail 'Shared libs not supported for Windows'
  }
}
```

Comment (boris-kolpackov) [reply build/root.build]: You can achieve this by overriding `bin.lib` in your `build/root.build`. For example, add something along these lines after loading the `cxx` module:

```
if ($cxx.target.class == 'windows')
{
  assert ($bin.lib != 'shared') 'only static variant can be built on Windows'
  bin.lib = static
}
```

For completeness you will also want to disable importing of the shared variant by modifying your `build/export.build` along these lines:

```
$out_root/
{
  include libhello/
}

if ($($out_root/libhello/ cxx.target.class) == 'windows')
  assert ($import.target != libs{hello}) "only static on Windows"

export $out_root/libhello/$import.target
```

I wouldn't recommend issuing a warning (there is nothing the user of your library can do about it), but if you want, you can modify the code in `build/root.build` along these lines:

```
if ($cxx.target.class == 'windows')
{
  switch $bin.lib
  {
    case 'shared'
      fail 'only static variant can be built on Windows'
    case 'both'
    {
      warn 'only static variant will be built on Windows'
      bin.lib = static
    }
  }
}
```

Comment (Swat-SomeBug) [reply build/root.build]: Cool!! Thanks @helmesjo and @boris-kolpackov, I never tried to set `bin.lib` within the `buildfile`! I’ll try it out and put it in!

Comment (helmesjo) [reply build/root.build]: @boris-kolpackov 
Is there anything additional that can be done to error out during importation against `google-benchmark%libs{benchmark}` (notice `libs{}`) when building `config.bin.lib=both`? Right now it will go through (even though there is no shared target) & instead fail during linking.

**Edit 1** I read a bit too fast, as that is what you are achieving with the edit in `export.build`. Wasn't able to do the same locally though, but it's probably just a typo somewhere on my end.

**Edit 2** Yep, typo. This solution would be perfect IMO.

File: build/root.build [hunk 2]
```diff
@@ -8,3 +8,18 @@ cxx{*}: extension = cc
...
 #
 test.target = $cxx.target
+
+# Disable support for building shared libs on windows with msvc
+# Upstream does not support this
+# Public headers are not marked appropriately with declspec(dllimport) and declspec(dllexport)
+# Causes severe problems with linking due to static variables and inline functions in headers
+if ($cxx.target.system == 'win32-msvc')
+{
+  switch $config.bin.lib
```
Comment (helmesjo) [inline build/root.build]: I think this should be only `$bin.lib`.

**Review by helmesjo — CHANGES_REQUESTED** [review]

File: build/root.build (hunk 2 above)
Comment (helmesjo) [reply build/root.build]: As I understand it, locally in a `buildfile`, eg. `$config.bin.lib` is the value passed from the build driver, and `$bin.lib` is the value for _this package_. The former is immutable, the latter is mutable. In this case, it probably doesn't matter, but it could be that `$config.bin.lib != $bin.lib` (in case somewhere in-between invoking the build driver & reaching `root.build` the latter value is modified, in a similar way to what we are doing here).
Probably nit-picking, but just `$bin.lib` seems more correct.

Comment (Swat-SomeBug) [reply build/root.build]: Yes, I also think `$config.bin.lib` is a value passed from the build driver and it immutable by packages. If this assumption is true, then wouldn't checking `$config.bin.lib` make more sense, since it is immutable, so no other package can change it. It also then capture's the users actual intent. `$bin.lib` is `google-benchmark`'s local "copy" that can only be changed by the package's buildfiles. @boris-kolpackov could you shed some light on the differences between `$config.bin.lib` and `$bin.lib`.
The change itself is ok to do. Just want to learn how these variables interact.

Comment (helmesjo) [reply build/root.build]: Yeah true, though _if_ `config.bin.lib` would differ to `bin.lib` (for whatever reason), this check could then do the "error out" decision incorrectly (eg. if it will actually be built as `static`, but we error out since the former value says `shared`).
But as you say, this scenario might not even be possible to achieve.

Comment (boris-kolpackov) [reply build/root.build]: Project-local `bin.lib` is initialized with a value from global `config.bin.lib` (or with a target platform-appropriate default). So checking/modifying `bin.lib` is the correct way to do it.

**Review by helmesjo — APPROVED** [review]
